pkg/util/leaderlease/leaderlease.go

```diff
@@ -0,0 +1,269 @@
+		return true, ttl, resp.EtcdIndex + 1, nil
+	}
+
+	if !storage.IsEtcdNodeExist(err) {
+		return false, 0, 0, fmt.Errorf("unable to check lease %s: %v", e.key, err)
+	}
+
+	latest, err := e.client.Get(e.key, false, false)
+	if err != nil {
+		return false, 0, 0, fmt.Errorf("unable to retrieve lease %s: %v", e.key, err)
+	}
+
+	index = eventIndexFor(latest)
+	if latest.Node.TTL > 0 {
+		ttl = uint64(latest.Node.TTL)
+	}
+
+	if latest.Node.Value != e.value {
+		glog.V(4).Infof("Lease %s owned by %s, waiting for expiration", e.key, latest.Node.Value)
+		// waits until the lease expires or changes to us
+		if err := e.waitForExpiration(false, &index, nil); err != nil {
+			return false, 0, 0, fmt.Errorf("unable to wait for lease expiration %s: %v", e.key, err)
+		}
+		return false, 0, 0, nil
+	}
+
+	glog.V(4).Infof("Lease %s already held, expires in %d seconds", e.key, ttl)
+
+	return true, ttl, index, nil
+}
+
+// Release tries to delete the leader lock.
+func (e *Etcd) Release() {
+	for i := 0; i < e.renewRetries; i++ {
+		_, err := e.client.CompareAndDelete(e.key, e.value, 0)
+		if err == nil {
+			break
+		}
+		if storage.IsEtcdTestFailed(err) || storage.IsEtcdNotFound(err) {
+			break
+		}
+		util.HandleError(fmt.Errorf("unable to release %s: %v", e.key, err))
+	}
+}
+
+// tryHold attempts to hold on to the lease by repeatedly refreshing its TTL.
+// If the lease hold fails, is deleted, or changed to another user
+func (e *Etcd) tryHold(ttl, index uint64) error {
+	// watch for termination
+	stop := make(chan struct{})
+	lost := make(chan struct{})
+	watchIndex := index
+	go util.Until(func() {
+		if err := e.waitForExpiration(true, &watchIndex, stop); err != nil {
+			util.HandleError(fmt.Errorf("error watching for lease expiration %s: %v", e.key, err))
+			return
+		}
+		glog.V(4).Infof("Lease %s lost due to deletion", e.key)
+		close(lost)
+	}, 100*time.Millisecond, stop)
+	defer close(stop)
+
+	duration := time.Duration(ttl) * time.Second
+	after := time.Duration(float32(duration) * e.waitFraction)
+	last := duration - after
+	interval := last / time.Duration(e.renewRetries)
+	if interval < e.minimumRetryInterval {
+		interval = e.minimumRetryInterval
+	}
+
+	// as long as we can renew the lease, loop
+	for {
+		select {
+		case <-time.After(after):
+			err := wait.Poll(interval, last, func() (bool, error) {
+				glog.V(4).Infof("Renewing lease %s", e.key)
+				resp, err := e.client.CompareAndSwap(e.key, e.value, e.ttl, e.value, index-1)
+				switch {
+				case err == nil:
+					index = eventIndexFor(resp)
+					return true, nil
+				case storage.IsEtcdTestFailed(err):
+					return false, fmt.Errorf("another client has taken the lease %s: %v", e.key, err)
+				case storage.IsEtcdNotFound(err):
+					return false, fmt.Errorf("another client has revoked the lease %s", e.key)
+				default:
+					util.HandleError(fmt.Errorf("unexpected error renewing lease %s: %v", e.key, err))
+					index = etcdIndexFor(err, index)
+					return false, nil
```

Comment:
We try again - if you return error it exits the Poll loop.  Will add a
comment.

On Fri, Aug 14, 2015 at 4:42 PM, David Eads <notifications@github.com>
wrote:

> In pkg/util/leaderlease/leaderlease.go
> <https://github.com/openshift/origin/pull/4137#discussion_r37117325>:
>
> > +		case <-time.After(after):
> > +			err := wait.Poll(interval, last, func() (bool, error) {
> > +				glog.V(4).Infof("Renewing lease %s", e.key)
> > +				resp, err := e.client.CompareAndSwap(e.key, e.value, e.ttl, e.value, index-1)
> > +				switch {
> > +				case err == nil:
> > +					index = eventIndexFor(resp)
> > +					return true, nil
> > +				case storage.IsEtcdTestFailed(err):
> > +					return false, fmt.Errorf("another client has taken the lease %s: %v", e.key, err)
> > +				case storage.IsEtcdNotFound(err):
> > +					return false, fmt.Errorf("another client has revoked the lease %s", e.key)
> > +				default:
> > +					util.HandleError(fmt.Errorf("unexpected error renewing lease %s: %v", e.key, err))
> > +					index = etcdIndexFor(err, index)
> > +					return false, nil
>
> Why do we return nil in this case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4137/files#r37117325>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift